File: app/src/main/java/com/medistock/util/TestUserSeeder.kt
package com.medistock.util

import com.medistock.shared.MedistockSDK
import com.medistock.shared.domain.model.Module
import com.medistock.shared.domain.model.User
import com.medistock.shared.domain.model.UserPermission
import java.util.UUID

/**
 * Utility class to seed test users for Maestro E2E permission tests.
 *
 * Usage: Call TestUserSeeder.seedTestUsers(sdk) during test setup or from a debug menu.
 *
 * All test users have the password: Test123!
 */
object TestUserSeeder {

    private const val TEST_PASSWORD = "Test123!"
    private const val SEEDER_ID = "test-seeder"

    /**
     * Test user configurations.
     * Each entry: username -> (fullName, isAdmin, list of modules with permissions)
     */
    private data class TestUserConfig(
        val username: String,
        val fullName: String,
        val isAdmin: Boolean = false,
        val permissions: Map<Module, PermissionFlags> = emptyMap()
    )

    private data class PermissionFlags(
        val canView: Boolean = false,
        val canCreate: Boolean = false,
        val canEdit: Boolean = false,
        val canDelete: Boolean = false
    ) {
        companion object {
            val FULL_CRUD = PermissionFlags(true, true, true, true)
            val VIEW_ONLY = PermissionFlags(true, false, false, false)
        }
    }

    private val testUsers = listOf(
        // No permissions user
        TestUserConfig(
            username = "test_no_permission",
            fullName = "Test No Permission"
        ),

        // Single module users - full CRUD
        TestUserConfig(
            username = "test_sites_only",
            fullName = "Test Sites Only",
            permissions = mapOf(Module.SITES to PermissionFlags.FULL_CRUD)
        ),
        TestUserConfig(
            username = "test_products_only",
            fullName = "Test Products Only",
            permissions = mapOf(Module.PRODUCTS to PermissionFlags.FULL_CRUD)
        ),
        TestUserConfig(
            username = "test_categories_only",
            fullName = "Test Categories Only",
            permissions = mapOf(Module.CATEGORIES to PermissionFlags.FULL_CRUD)
        ),
        TestUserConfig(
            username = "test_customers_only",
            fullName = "Test Customers Only",
            permissions = mapOf(Module.CUSTOMERS to PermissionFlags.FULL_CRUD)
        ),
        TestUserConfig(
            username = "test_packaging_only",
            fullName = "Test Packaging Only",
            permissions = mapOf(Module.PACKAGING_TYPES to PermissionFlags.FULL_CRUD)
        ),
        TestUserConfig(
            username = "test_users_only",
            fullName = "Test Users Only",
            permissions = mapOf(Module.USERS to PermissionFlags.FULL_CRUD)
        ),

        // Single module users - view only
        TestUserConfig(
            username = "test_stock_only",
            fullName = "Test Stock View Only",
            permissions = mapOf(Module.STOCK to PermissionFlags.VIEW_ONLY)
        ),
        TestUserConfig(
            username = "test_audit_only",
            fullName = "Test Audit View Only",
            permissions = mapOf(Module.AUDIT to PermissionFlags.VIEW_ONLY)
        ),

        // Operation users - full CRUD
        TestUserConfig(
            username = "test_purchases_only",
            fullName = "Test Purchases Only",
            permissions = mapOf(Module.PURCHASES to PermissionFlags.FULL_CRUD)
        ),
        TestUserConfig(
            username = "test_sales_only",
            fullName = "Test Sales Only",
            permissions = mapOf(Module.SALES to PermissionFlags.FULL_CRUD)
        ),
        TestUserConfig(
            username = "test_transfers_only",
            fullName = "Test Transfers Only",
            permissions = mapOf(Module.TRANSFERS to PermissionFlags.FULL_CRUD)
        ),
        TestUserConfig(
            username = "test_inventory_only",
            fullName = "Test Inventory Only",
            permissions = mapOf(Module.INVENTORY to PermissionFlags.FULL_CRUD)
        ),

        // CRUD granularity users (for Products module)
        TestUserConfig(
            username = "test_products_view",
            fullName = "Test Products View Only",
            permissions = mapOf(Module.PRODUCTS to PermissionFlags.VIEW_ONLY)
        ),
        TestUserConfig(
            username = "test_products_create",
            fullName = "Test Products Create",
            permissions = mapOf(Module.PRODUCTS to PermissionFlags(true, true, false, false))
        ),
        TestUserConfig(
            username = "test_products_edit",
            fullName = "Test Products Edit",
            permissions = mapOf(Module.PRODUCTS to PermissionFlags(true, false, true, false))
        ),
        TestUserConfig(
            username = "test_products_delete",
            fullName = "Test Products Delete",
            permissions = mapOf(Module.PRODUCTS to PermissionFlags(true, false, false, true))
        ),

        // Multi-permission user
        TestUserConfig(
            username = "test_multi_perm",
            fullName = "Test Multi Permission",
            permissions = mapOf(
                Module.SITES to PermissionFlags.FULL_CRUD,
                Module.PRODUCTS to PermissionFlags.FULL_CRUD,
                Module.SALES to PermissionFlags.FULL_CRUD
            )
        ),

        // Admin user (for reference)
        TestUserConfig(
            username = "test_admin",
            fullName = "Test Admin",
            isAdmin = true
        )
    )

    /**
     * Seed all test users into the database.
     * Existing users with the same username will be skipped.
     *
     * @param sdk The MedistockSDK instance
     * @return Number of users created
     */
    suspend fun seedTestUsers(sdk: MedistockSDK): Int {
        val currentTime = System.currentTimeMillis()
        val hashedPassword = PasswordHasher.hashPassword(TEST_PASSWORD)
        var createdCount = 0

        for (config in testUsers) {
            // Check if user already exists
            val existingUser = sdk.userRepository.getByUsername(config.username)
            if (existingUser != null) {
                println("TestUserSeeder: User ${config.username} already exists, skipping")
                continue
            }

            // Create user
            val userId = UUID.randomUUID().toString()
            val user = User(
                id = userId,
                username = config.username,
                password = hashedPassword,
                fullName = config.fullName,
                isAdmin = config.isAdmin,
                isActive = true,
                createdAt = currentTime,
                updatedAt = currentTime,
                createdBy = SEEDER_ID,
                updatedBy = SEEDER_ID
            )

            sdk.userRepository.insert(user)
            println("TestUserSeeder: Created user ${config.username}")

            // Create permissions (only for non-admin users)
            if (!config.isAdmin) {
                for ((module, flags) in config.permissions) {
                    val permission = UserPermission(
                        id = UUID.randomUUID().toString(),
                        userId = userId,
                        module = module.name,
                        canView = flags.canView,
                        canCreate = flags.canCreate,
                        canEdit = flags.canEdit,
                        canDelete = flags.canDelete,
                        createdAt = currentTime,
                        updatedAt = currentTime,
                        createdBy = SEEDER_ID,
                        updatedBy = SEEDER_ID
                    )
                    sdk.userPermissionRepository.insert(permission)
                }
                println("TestUserSeeder: Created ${config.permissions.size} permissions for ${config.username}")
            }

            createdCount++
        }

        println("TestUserSeeder: Seeding complete. Created $createdCount users.")
        return createdCount
    }

    /**
     * Remove all test users from the database.
     * Only removes users created by the seeder (identified by username prefix).
     *
     * @param sdk The MedistockSDK instance
     * @return Number of users removed
     */
    suspend fun removeTestUsers(sdk: MedistockSDK): Int {
        var removedCount = 0

        for (config in testUsers) {
            val existingUser = sdk.userRepository.getByUsername(config.username)
            if (existingUser != null) {
                // Delete permissions first
                sdk.userPermissionRepository.deletePermissionsForUser(existingUser.id)
                // Delete user
                sdk.userRepository.delete(existingUser.id)
                println("TestUserSeeder: Removed user ${config.username}")
                removedCount++
            }
        }

        println("TestUserSeeder: Cleanup complete. Removed $removedCount users.")
        return removedCount
    }

    /**
     * Get the list of test usernames.
     */
    fun getTestUsernames(): List<String> = testUsers.map { it.username }

    /**
     * Get the test password (for documentation/test scripts).
     */
    fun getTestPassword(): String = TEST_PASSWORD
}

File: iosApp/iosApp/Services/TestUserSeeder.swift
import Foundation
import shared

/// Utility class to seed test users for Maestro E2E permission tests.
///
/// Usage: Call TestUserSeeder.seedTestUsers(sdk:) during test setup.
///
/// All test users have the password: Test123!
final class TestUserSeeder {

    static let instance = TestUserSeeder()

    private let testPassword = "Test123!"
    private let seederId = "test-seeder"

    private init() {}

    // MARK: - Test User Configurations

    private struct TestUserConfig {
        let username: String
        let fullName: String
        let isAdmin: Bool
        let permissions: [String: PermissionFlags]

        init(username: String, fullName: String, isAdmin: Bool = false, permissions: [String: PermissionFlags] = [:]) {
            self.username = username
            self.fullName = fullName
            self.isAdmin = isAdmin
            self.permissions = permissions
        }
    }

    private struct PermissionFlags {
        let canView: Bool
        let canCreate: Bool
        let canEdit: Bool
        let canDelete: Bool

        static let fullCrud = PermissionFlags(canView: true, canCreate: true, canEdit: true, canDelete: true)
        static let viewOnly = PermissionFlags(canView: true, canCreate: false, canEdit: false, canDelete: false)
    }

    private let testUsers: [TestUserConfig] = [
        // No permissions user
        TestUserConfig(username: "test_no_permission", fullName: "Test No Permission"),

        // Single module users - full CRUD
        TestUserConfig(username: "test_sites_only", fullName: "Test Sites Only", permissions: ["SITES": .fullCrud]),
        TestUserConfig(username: "test_products_only", fullName: "Test Products Only", permissions: ["PRODUCTS": .fullCrud]),
        TestUserConfig(username: "test_categories_only", fullName: "Test Categories Only", permissions: ["CATEGORIES": .fullCrud]),
        TestUserConfig(username: "test_customers_only", fullName: "Test Customers Only", permissions: ["CUSTOMERS": .fullCrud]),
        TestUserConfig(username: "test_packaging_only", fullName: "Test Packaging Only", permissions: ["PACKAGING_TYPES": .fullCrud]),
        TestUserConfig(username: "test_users_only", fullName: "Test Users Only", permissions: ["USERS": .fullCrud]),

        // Single module users - view only
        TestUserConfig(username: "test_stock_only", fullName: "Test Stock View Only", permissions: ["STOCK": .viewOnly]),
        TestUserConfig(username: "test_audit_only", fullName: "Test Audit View Only", permissions: ["AUDIT": .viewOnly]),

        // Operation users - full CRUD
        TestUserConfig(username: "test_purchases_only", fullName: "Test Purchases Only", permissions: ["PURCHASES": .fullCrud]),
        TestUserConfig(username: "test_sales_only", fullName: "Test Sales Only", permissions: ["SALES": .fullCrud]),
        TestUserConfig(username: "test_transfers_only", fullName: "Test Transfers Only", permissions: ["TRANSFERS": .fullCrud]),
        TestUserConfig(username: "test_inventory_only", fullName: "Test Inventory Only", permissions: ["INVENTORY": .fullCrud]),

        // CRUD granularity users (for Products module)
        TestUserConfig(username: "test_products_view", fullName: "Test Products View Only", permissions: ["PRODUCTS": .viewOnly]),
        TestUserConfig(username: "test_products_create", fullName: "Test Products Create", permissions: ["PRODUCTS": PermissionFlags(canView: true, canCreate: true, canEdit: false, canDelete: false)]),
        TestUserConfig(username: "test_products_edit", fullName: "Test Products Edit", permissions: ["PRODUCTS": PermissionFlags(canView: true, canCreate: false, canEdit: true, canDelete: false)]),
        TestUserConfig(username: "test_products_delete", fullName: "Test Products Delete", permissions: ["PRODUCTS": PermissionFlags(canView: true, canCreate: false, canEdit: false, canDelete: true)]),

        // Multi-permission user
        TestUserConfig(username: "test_multi_perm", fullName: "Test Multi Permission", permissions: [
            "SITES": .fullCrud,
            "PRODUCTS": .fullCrud,
            "SALES": .fullCrud
        ]),

        // Admin user (for reference)
        TestUserConfig(username: "test_admin", fullName: "Test Admin", isAdmin: true)
    ]

    // MARK: - Public Methods

    /// Seed all test users into the database.
    /// Existing users with the same username will be skipped.
    ///
    /// - Parameter sdk: The MedistockSDK instance
    /// - Returns: Number of users created
    func seedTestUsers(sdk: MedistockSDK) async -> Int {
        let currentTime = Int64(Date().timeIntervalSince1970 * 1000)
        guard let hashedPassword = PasswordHasher.shared.hashPassword(testPassword) else {
            print("TestUserSeeder: Failed to hash password")
            return 0
        }

        var createdCount = 0

        for config in testUsers {
            do {
                // Check if user already exists
                let existingUser = try await sdk.userRepository.getByUsername(username: config.username)
                if existingUser != nil {
                    print("TestUserSeeder: User \(config.username) already exists, skipping")
                    continue
                }

                // Create user
                let userId = UUID().uuidString
                let user = User(
                    id: userId,
                    username: config.username,
                    password: hashedPassword,
                    fullName: config.fullName,
                    language: nil,
                    isAdmin: config.isAdmin,
                    isActive: true,
                    createdAt: currentTime,
                    updatedAt: currentTime,
                    createdBy: seederId,
                    updatedBy: seederId
                )

                try await sdk.userRepository.insert(user: user)
                print("TestUserSeeder: Created user \(config.username)")

                // Create permissions (only for non-admin users)
                if !config.isAdmin {
                    for (module, flags) in config.permissions {
                        let permission = shared.UserPermission(
                            id: UUID().uuidString,
                            userId: userId,
                            module: module,
                            canView: flags.canView,
                            canCreate: flags.canCreate,
                            canEdit: flags.canEdit,
                            canDelete: flags.canDelete,
                            createdAt: currentTime,
                            updatedAt: currentTime,
                            createdBy: seederId,
                            updatedBy: seederId
                        )
                        try await sdk.userPermissionRepository.insert(permission: permission)
                    }
                    print("TestUserSeeder: Created \(config.permissions.count) permissions for \(config.username)")
                }

                createdCount += 1
            } catch {
                print("TestUserSeeder: Error creating user \(config.username): \(error)")
            }
        }

        print("TestUserSeeder: Seeding complete. Created \(createdCount) users.")
        return createdCount
    }

    /// Remove all test users from the database.
    /// Only removes users created by the seeder (identified by username prefix).
    ///
    /// - Parameter sdk: The MedistockSDK instance
    /// - Returns: Number of users removed
    func removeTestUsers(sdk: MedistockSDK) async -> Int {
        var removedCount = 0

        for config in testUsers {
            do {
                let existingUser = try await sdk.userRepository.getByUsername(username: config.username)
                if let user = existingUser {
                    // Delete permissions first
                    try await sdk.userPermissionRepository.deletePermissionsForUser(userId: user.id)
                    // Delete user
                    try await sdk.userRepository.delete(id: user.id)
                    print("TestUserSeeder: Removed user \(config.username)")
                    removedCount += 1
                }
            } catch {
                print("TestUserSeeder: Error removing user \(config.username): \(error)")
            }
        }

        print("TestUserSeeder: Cleanup complete. Removed \(removedCount) users.")
        return removedCount
    }

    /// Get the list of test usernames.
    func getTestUsernames() -> [String] {
        testUsers.map { $0.username }
    }

    /// Get the test password (for documentation/test scripts).
    func getTestPassword() -> String {
        testPassword
    }
}
